### ISSUE TITLE
Max vcpu limit of 255 is too low

### ISSUE THREAD

**Issue by peng6662001** [cloud-hypervisor/cloud-hypervisor#5345]
**Describe the bug**
The VM on Ampere Altra returns `Error parsing config: Error parsing --cpus: unable to convert 256 for boot` when I use `--cpus boot=256,max=256`.
I found the root cuase is that the size of boot_vcpus is too small,please help to fix it.
```
vmm/src/config.rs
CpusConfig
  fn parse
    let boot_vcpus: u8 = parser
            .convert("boot")
            .map_err(Error::ParseCpus)?
            .unwrap_or(DEFAULT_VCPUS);
```

**VM configuration**
```
target/aarch64-unknown-linux-gnu/debug/cloud-hypervisor --cpus boot=256,max=256 --memory size=4G --kernel /home/dom/cloud-hypervisor/bin/Image --cmdline 'root=/dev/vda1 rw systemd.journald.forward_to_console=1' --disk path=/root/osdisk.img --disk path=/root/cloudinit --net tap=,mac=12:34:56:78:90:00,ip=172.19.0.1,mask=255.255.255.0,num_queues=2,queue_size=256
```

Guest OS version details:
cloud-hypervisor v30.0-93-gd3ac6a85

Host OS version details:
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.1 LTS
Release:        22.04
Codename:       jammy

**Comment by MrXinWang** [comment]
By taking a quick glance, I don't really think you can assign a value bigger than 255 to a `u8`, so I think this is a common issue instead of an Arm-specific issue.

**Comment by peng6662001** [comment]
@MrXinWang How about changing the size of boot_vcpus to u16.

**Comment by MrXinWang** [comment]
I do have the same question in my mind. But I am not 100% sure if this is intentional.

Hi @rbradford @likebreath, apology for this sudden question, is there any blocker in your mind to change this field to something larger than `u8`? Do you remember if there were some specific thoughts about this field when designing the API?

**Comment by peng6662001** [comment]
Ampere's new server has 320 cpu, is cloud hypervisor willing to support more than 255 cpu?

**Comment by likebreath** [comment]
Sorry about missing this issue. I believe the we started with `u8` as it was simply enough at the time. 

The hypervisor (kvm or mshv) has a limit on the maximum number of vcpus it supports, and we have a check on that here: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/7f18d0a281a3c03e9c9787087bcace487b2b7e32/vmm/src/cpu.rs#L609-L611

So I think it should be safe to bump the vcpu related parameters from `u8` to `u16` from CH POV. But I am not sure how useful this would be in practice. Also, I want to mention that this change would break live-upgrades and live-migration cross CH versions.

@rbradford @liuw Any thoughts?

**Comment by liuw** [comment]
> Sorry about missing this issue. I believe the we started with `u8` as it was simply enough at the time.
> 
> The hypervisor (kvm or mshv) has a limit on the maximum number of vcpus it supports, and we have a check on that here:
> 
> https://github.com/cloud-hypervisor/cloud-hypervisor/blob/7f18d0a281a3c03e9c9787087bcace487b2b7e32/vmm/src/cpu.rs#L609-L611
> 
> So I think it should be safe to bump the vcpu related parameters from `u8` to `u16` from CH POV. But I am not sure how useful this would be in practice. Also, I want to mention that this change would break live-upgrades and live-migration cross CH versions.
> 
> @rbradford @liuw Any thoughts?

How does this break cross version migration? Are you worried about migrating from N to N-1?

**Comment by likebreath** [comment]
> > Sorry about missing this issue. I believe the we started with `u8` as it was simply enough at the time.
> > The hypervisor (kvm or mshv) has a limit on the maximum number of vcpus it supports, and we have a check on that here:
> > https://github.com/cloud-hypervisor/cloud-hypervisor/blob/7f18d0a281a3c03e9c9787087bcace487b2b7e32/vmm/src/cpu.rs#L609-L611
> > 
> > So I think it should be safe to bump the vcpu related parameters from `u8` to `u16` from CH POV. But I am not sure how useful this would be in practice. Also, I want to mention that this change would break live-upgrades and live-migration cross CH versions.
> > @rbradford @liuw Any thoughts?
> 
> How does this break cross version migration? Are you worried about migrating from N to N-1?

No, I think upgrade or migration to newer versions will break, as data type change will break serialization. For this case it will break serialization on `CpusConfig`.

**Comment by liuw** [comment]
> > > Sorry about missing this issue. I believe the we started with `u8` as it was simply enough at the time.
> > > The hypervisor (kvm or mshv) has a limit on the maximum number of vcpus it supports, and we have a check on that here:
> > > https://github.com/cloud-hypervisor/cloud-hypervisor/blob/7f18d0a281a3c03e9c9787087bcace487b2b7e32/vmm/src/cpu.rs#L609-L611
> > > 
> > > So I think it should be safe to bump the vcpu related parameters from `u8` to `u16` from CH POV. But I am not sure how useful this would be in practice. Also, I want to mention that this change would break live-upgrades and live-migration cross CH versions.
> > > @rbradford @liuw Any thoughts?
> > 
> > 
> > How does this break cross version migration? Are you worried about migrating from N to N-1?
> 
> No, I think upgrade or migration to newer versions will break, as data type change will break serialization. For this case it will break serialization on `CpusConfig`.

Hmm... I thought u16 is strictly larger than u8 so the conversion should be fine, but I didn't check the code.

I _think_ the data is encoded in JSON which does not contain type information for integer, right?

**Comment by likebreath** [comment]
> > No, I think upgrade or migration to newer versions will break, as data type change will break serialization. For this case it will break serialization on `CpusConfig`.
> 
> Hmm... I thought u16 is strictly larger than u8 so the conversion should be fine, but I didn't check the code.
> 
> I _think_ the data is encoded in JSON which does not contain type information for integer, right?
> 

I assumed that serialization data would be changed if data type changed (which is true for serialization in c++ using boost binary serialization). You might be right here as JSON is loosely typed. 

We have live-upgrade tests enabled in our CI, so it will tell us whether it is a problem. :)

> In any case, I think 255 is too low. Some cloud providers have VMs larger than that, so it will need to be changed.

That's good to know. For this case, I agree it will need to be changed.

**Comment by rbradford** [comment]
Bumping the size of the vCPUs (max/boot) in the config is safe as the config is serialised to/from JSON which has no idea about the size of types.

Unfortunately there is a lot more to it than just changing this one field - you can see some of the work that I started on this in this PR - https://github.com/cloud-hypervisor/cloud-hypervisor/pull/5600

**Comment by liuw** [comment]
For x86, CH will need to use X2APIC mode. Not sure what's needed for ARM64 -- that seems to be what @peng6662001 is working on.

**Comment by KshitijKapoor8** [comment]
Hey, my name is Kshitij and I am a student at UT Austin. I was wondering if this issue still needs to be addressed and is relevant, and if so, could I be assigned to try to tackle this (probably branching off of the work that @rbradford did)?

**Comment by rbradford** [comment]
#7299 increased the number of supported CPUs on x86-64 to 8192 so i'm going to update this to be aarch64 specific
